### ISSUE TITLE
Explicit path alias matching order for coverage combine

### ISSUE THREAD

**Issue by nedbat** [coveragepy/coveragepy#649]
*Originally reported by* **Jonathan Beezley (Bitbucket: [jbeezley](https://bitbucket.org/jbeezley), GitHub: [jbeezley](https://github.com/jbeezley))**

----------------------------------------

Over in https://github.com/girder/girder, we have a fairly convoluted use case where we need to use path aliases for our `coverage combine` step, but we also need ensure that the matching occurs in the correct order.  To gloss over some of the messy details, the issue is that in a development install we have the following directory structure:
```
root/ ->
    girder/
        __init__.py
        ...
    plugins/
        __init__.py
        ...
```
But when installed, the package looks like this:
```
root/ ->
    girder/
        __init__.py
        ...
        plugins/
            __init__.py
            ...
```

We need to be able to combine coverage files executed from both of these environments,so we use a path alias that looks like this:
```
[coverage:paths]
plugins =
    plugins/
    build/test/tox/*/lib/*/site-packages/girder/plugins/
girder = 
    girder/
    build/test/tox/*/lib/*/site-packages/girder/
```
The issue here is that it matches the `girder` section first and aliases the plugin modules to `$PWD/girder/plugins` rather than `$PWD/plugins`.

Globbing patterns don't allow us to exclude only certain subdirectories from the second group, so we need to have some way to ensure that the it is matched after the first.

I would be willing to implement one of the following solutions.  I would appreciate any guidance as to which approach (if any) would be preferred by the maintainers.

1. Somehow allow explicit regex's in the path matching.  I'm not sure how the syntax of this would work.
2. Use an `OrderedDict` on [this line](https://bitbucket.org/ned/coveragepy/src/acaeb69b6a7147c872dd9261bca96fa2b52260c6/coverage/config.py#config.py-222) so that the original order in the config file is preserved.  This would be ideal, but it would require a backport of `OrderedDict` for python 2.6 support.
3. Iterate through the keys in sorted order on [this line](https://bitbucket.org/ned/coveragepy/src/acaeb69b6a7147c872dd9261bca96fa2b52260c6/coverage/files.py#files.py-380).  Users could then choose to name the alias targets so they apply in the required order.  This is the simplest solution that I can think of, but it smells a little hacky to me.

----------------------------------------
- Bitbucket: https://bitbucket.org/ned/coveragepy/issue/649

**Comment by nedbat** [comment]
@jbeezley An OrderedDict sounds like a good approach.  I've dropped Python 2.6 support since you wrote this up.

**Comment by nedbat** [comment]
I've fixed this in 65b9007b.  Thanks for the detailed example. It's now enshrined in my test suite!

**Comment by nedbat** [comment]
This was released as part of 5.0b2 today.

**Comment by nedbat** [comment]
This is now improved in commit 7e7c44b4f5c484105559690b1efccd84839bc640.  The result of a remapping must exist, so now this scenario will find the correct file with the path aliases in either order.

**Comment by nedbat** [comment]
This is now released as part of [coverage 7.0.0b1](https://pypi.org/project/coverage/7.0.0b1).
